tests/e2e_tests/test_pipeline.py

```diff
@@ -17,7 +17,7 @@
     depends=[
         'e2e_tests/test_subscribe_handler.py::test_subscribe_node_channel'],
     scope='session')
-@pytest.mark.order(5)
+@pytest.mark.order(4)
```

Comment:
Yes, it is cumbersome to maintain in the long run. But I don't think there is much we can do about it.
Maybe using `before` and `after` marker attributes will optimize this a bit. I'll try and test it.

Reply:
OK. So, I tested with [markers](https://pytest-order.readthedocs.io/en/stable/usage.html#order-relative-to-other-tests) used for reletive tests. However, it doesn't work as expected and multiple tests are being skipped because of the tests dependencies.
```
kernelci-api-e2e-tests | e2e_tests/test_user_creation.py::test_create_admin_user PASSED           [  5%]
kernelci-api-e2e-tests | e2e_tests/test_count_handler.py::test_count_nodes SKIPPED (test_coun...) [ 11%]
kernelci-api-e2e-tests | e2e_tests/test_count_handler.py::test_count_nodes_matching_attributes SKIPPED [ 17%]
kernelci-api-e2e-tests | e2e_tests/test_pipeline.py::test_node_pipeline SKIPPED (test_node_pi...) [ 23%]
kernelci-api-e2e-tests | e2e_tests/test_pubsub_handler.py::test_pubsub_handler SKIPPED (test_...) [ 29%]
kernelci-api-e2e-tests | e2e_tests/test_regression_handler.py::test_regression_handler SKIPPED    [ 35%]
kernelci-api-e2e-tests | e2e_tests/test_root_handler.py::test_root_endpoint PASSED                [ 41%]
kernelci-api-e2e-tests | e2e_tests/test_user_creation.py::test_create_regular_user PASSED         [ 47%]
kernelci-api-e2e-tests | e2e_tests/test_subscribe_handler.py::test_subscribe_node_channel PASSED  [ 52%]
kernelci-api-e2e-tests | e2e_tests/test_subscribe_handler.py::test_subscribe_test_channel PASSED  [ 58%]
kernelci-api-e2e-tests | e2e_tests/test_subscribe_handler.py::test_subscribe_user_group_channel PASSED [ 64%]
kernelci-api-e2e-tests | e2e_tests/test_user_creation.py::test_whoami PASSED                      [ 70%]
kernelci-api-e2e-tests | e2e_tests/test_user_creation.py::test_create_user_negative PASSED        [ 76%]
kernelci-api-e2e-tests | e2e_tests/test_user_group_handler.py::test_create_and_get_user_group PASSED [ 82%]
kernelci-api-e2e-tests | e2e_tests/test_password_handler.py::test_password_endpoint PASSED        [ 88%]
kernelci-api-e2e-tests | e2e_tests/test_unsubscribe_handler.py::test_unsubscribe_node_channel PASSED [ 94%]
kernelci-api-e2e-tests | e2e_tests/test_unsubscribe_handler.py::test_unsubscribe_test_channel PASSED [100%]
kernelci-api-e2e-tests | ======== 12 passed, 5 skipped, 13 warnings in 1.19s==================
```
So, I guess ordering with numbers is the only way here.

Reply:
Yes, dependencies are defined by `pytest-dependency`. In some cases, we need to execute tests in a specific sequence. For instance, we need to create users before sending authenticated requests. So, instead of adding the dependency marker to all the authenticated tests, we can add order marker to make sure the test creating user executes before any another tests.

Reply:
The point of using order markers is, that if you just use dependencies, it does not guarantee that a test will only execute after its dependency. Tests will execute in any order and see if the dependency is `PASSED` otherwise the test will be `SKIPPED`. It does not make sure that dependencies will execute before the dependent. That's when `order` comes into the picture.

For example, when I removed `order` from all the tests except user and user group creation, this is what I got:
```
kernelci-api-e2e-tests | 
kernelci-api-e2e-tests | e2e_tests/test_user_group_handler.py::test_create_user_groups PASSED     [  5%]
kernelci-api-e2e-tests | e2e_tests/test_user_creation.py::test_create_admin_user PASSED           [ 10%]
kernelci-api-e2e-tests | e2e_tests/test_user_creation.py::test_create_regular_user PASSED         [ 15%]
kernelci-api-e2e-tests | e2e_tests/test_count_handler.py::test_count_nodes SKIPPED (test_coun...) [ 21%]
kernelci-api-e2e-tests | e2e_tests/test_count_handler.py::test_count_nodes_matching_attributes SKIPPED [ 26%]
kernelci-api-e2e-tests | e2e_tests/test_password_handler.py::test_password_endpoint PASSED        [ 31%]
kernelci-api-e2e-tests | e2e_tests/test_pipeline.py::test_node_pipeline SKIPPED (test_node_pi...) [ 36%]
kernelci-api-e2e-tests | e2e_tests/test_pubsub_handler.py::test_pubsub_handler SKIPPED (test_...) [ 42%]
kernelci-api-e2e-tests | e2e_tests/test_regression_handler.py::test_regression_handler SKIPPED    [ 47%]
kernelci-api-e2e-tests | e2e_tests/test_root_handler.py::test_root_endpoint PASSED                [ 52%]
kernelci-api-e2e-tests | e2e_tests/test_subscribe_handler.py::test_subscribe_node_channel PASSED  [ 57%]
kernelci-api-e2e-tests | e2e_tests/test_subscribe_handler.py::test_subscribe_test_channel PASSED  [ 63%]
kernelci-api-e2e-tests | e2e_tests/test_subscribe_handler.py::test_subscribe_user_group_channel PASSED [ 68%]
kernelci-api-e2e-tests | e2e_tests/test_unsubscribe_handler.py::test_unsubscribe_node_channel PASSED [ 73%]
kernelci-api-e2e-tests | e2e_tests/test_unsubscribe_handler.py::test_unsubscribe_test_channel PASSED [ 78%]
kernelci-api-e2e-tests | e2e_tests/test_user_creation.py::test_whoami PASSED                      [ 84%]
kernelci-api-e2e-tests | e2e_tests/test_user_creation.py::test_create_user_negative PASSED        [ 89%]
kernelci-api-e2e-tests | e2e_tests/test_user_group_handler.py::test_get_user_group PASSED         [ 94%]
kernelci-api-e2e-tests | e2e_tests/test_user_group_handler.py::test_create_and_get_user_group PASSED [100%]
kernelci-api-e2e-tests | e2e_tests/test_user_group_handler.py::test_create_and_get_user_group ERROR [100%]
kernelci-api-e2e-tests | 
kernelci-api-e2e-tests | ============= 14 passed, 5 skipped, 13 warnings, 1 error in 1.32s ====
```

Still, if you feel that you can come up with a better solution, that'd be welcome.